File: app/src/main/java/de/hsesslingen/stundenplan/data/NextEvent.kt
package de.hsesslingen.stundenplan.data

import java.time.LocalDate
import java.time.LocalDateTime

data class NextEventResult(val event: TimetableEvent, val date: LocalDate)

/**
 * Finds the next upcoming event at or after [now] within [events] (which should cover at least
 * [now]'s week and the following one, so a search starting on a Friday evening can still find
 * Monday's first lecture). Weekends are skipped since QIS never schedules on them. An event
 * currently in progress still counts as "next" — only ones that have fully ended are skipped.
 */
fun findNextEvent(
    events: List<TimetableEvent>,
    now: LocalDateTime,
    hiddenGroupKeys: Set<String> = emptySet(),
    daysAhead: Int = 7,
): NextEventResult? = findUpcomingEvents(events, now, hiddenGroupKeys, count = 1, daysAhead = daysAhead).firstOrNull()

/** Like [findNextEvent] but returns up to [count] events in chronological order — used by the
 *  home-screen widget's larger size variant, which has room to show more than just the next one. */
fun findUpcomingEvents(
    events: List<TimetableEvent>,
    now: LocalDateTime,
    hiddenGroupKeys: Set<String> = emptySet(),
    count: Int = 3,
    daysAhead: Int = 7,
): List<NextEventResult> {
    val nowMinutes = now.hour * 60 + now.minute
    val result = mutableListOf<NextEventResult>()
    for (offset in 0..daysAhead) {
        if (result.size >= count) break
        val date = now.toLocalDate().plusDays(offset.toLong())
        val weekday = Weekday.fromDate(date) ?: continue
        val dayEvents = events
            .filter { it.day == weekday && it.groupKey !in hiddenGroupKeys }
            .filter { offset > 0 || it.endMinutes > nowMinutes }
            .sortedBy { it.startMinutes }
        for (event in dayEvents) {
            result += NextEventResult(event, date)
            if (result.size >= count) break
        }
    }
    return result
}

File: desktop/src/core/NextEvent.cpp
#include "NextEvent.h"

#include <algorithm>

namespace stundenplan {

QList<NextEventResult> findUpcomingEvents(const QList<TimetableEvent> &events,
                                           const QDateTime &now,
                                           const QSet<QString> &hiddenGroupKeys,
                                           int count,
                                           int daysAhead)
{
    const int nowMinutes = now.time().hour() * 60 + now.time().minute();
    QList<NextEventResult> result;

    for (int offset = 0; offset <= daysAhead; ++offset) {
        if (result.size() >= count)
            break;
        const QDate date = now.date().addDays(offset);
        auto weekdayOpt = weekdayFromDate(date);
        if (!weekdayOpt.has_value())
            continue;
        const Weekday weekday = *weekdayOpt;

        QList<TimetableEvent> dayEvents;
        for (const auto &event : events) {
            if (event.day != weekday)
                continue;
            if (hiddenGroupKeys.contains(event.groupKey()))
                continue;
            if (!(offset > 0 || event.endMinutes > nowMinutes))
                continue;
            dayEvents.append(event);
        }
        std::sort(dayEvents.begin(), dayEvents.end(), [](const TimetableEvent &a, const TimetableEvent &b) {
            return a.startMinutes < b.startMinutes;
        });

        for (const auto &event : dayEvents) {
            result.append(NextEventResult{event, date});
            if (result.size() >= count)
                break;
        }
    }
    return result;
}

std::optional<NextEventResult> findNextEvent(const QList<TimetableEvent> &events,
                                              const QDateTime &now,
                                              const QSet<QString> &hiddenGroupKeys,
                                              int daysAhead)
{
    auto results = findUpcomingEvents(events, now, hiddenGroupKeys, 1, daysAhead);
    if (results.isEmpty())
        return std::nullopt;
    return results.first();
}

} // namespace stundenplan
